### PR TITLE
[DI] Check for privates before shared services

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 3.2
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | https://github.com/symfony/symfony/pull/22801#discussion_r117732213, https://github.com/symfony/symfony/pull/22801#discussion_r117732599
| License       | MIT
| Doc PR        | symfony/symfony-docs#... 

cc @stof

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
CI is red :)

**Comment by stof** [comment]
@nicolas-grekas this is because of https://github.com/composer/composer/issues/6342

**Comment by ro0NL** [comment]
Kinda clueless here... trying to fix `WebProfilerExtensionTest::testToolbarConfig`

patched `assertSaneContainer` so it skips private id's (hardcoded locally to test :)

however these 2 assertions pass for all provided debug modes, _indiviudally_;

```php
        $this->assertSaneContainer($this->getDumpedContainer(), '', array('web_profiler.csp.handler'));

        if ($listenerInjected) {
            $this->assertSame($listenerEnabled, $this->container->get('web_profiler.debug_toolbar')->isEnabled());
        }
```

the combination seems to trigger a deprecation :confused:

**Comment by ro0NL** [comment]
first patch.. lets see what happens :)

**Comment by ro0NL** [comment]
`testStopwatchExtensionAvailability` actually gets a private service :angry: 

`$container->get('twig.extension.debug.stopwatch')->getTokenParsers()`

is there a clean way to silence expected deprecations.. or should we do something else here.

**Comment by stof** [comment]
well, the test should either not access a private service, or be marked as legacy. If the test is not meant to be removed in 4.0, marking it as legacy is not an option, and the test should be rewritten instead.
This is good that I catched the mistake in the deprecation trigger in the other PR, as this allowed us to catch mistakes in our own testsuite which was using deprecated APIs without getting the notice...

**Comment by ro0NL** [comment]
Updated.. this should do. But i dont like it really :) but i guess we have no choice as we never decided how privates reflects a compiled containerbuilder.

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php [hunk 1]
```diff
@@ -208,9 +208,13 @@ public function testStopwatchExtensionAvailability($debug, $stopwatchEnabled, $e
...
         $container->registerExtension(new TwigExtension());
         $container->loadFromExtension('twig', array());
+
+        $container->register('test_debug_stopwatch', 'stdClass')
+            ->setProperty('stopwatch', new Reference('twig.extension.debug.stopwatch'));
+
         $this->compileContainer($container);
 
-        $tokenParsers = $container->get('twig.extension.debug.stopwatch')->getTokenParsers();
+        $tokenParsers = $container->get('test_debug_stopwatch')->stopwatch->getTokenParsers();
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: a better solution would be to use `$container->get('twig')->getExtension(StopwatchExtension::class)` to access the extension, which avoids registering an extra service

File: src/Symfony/Component/DependencyInjection/Container.php [hunk 2]
```diff
@@ -262,6 +262,10 @@ public function get($id, $invalidBehavior = self::EXCEPTION_ON_INVALID_REFERENCE
             if (isset($this->aliases[$id])) {
                 $id = $this->aliases[$id];
             }
+            if (isset($this->privates[$id]) && !$this instanceof ContainerBuilder) {
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Container.php]: why the check on ContainerBuilder ? this is useless (`$this->privates` will be empty for them anyway)

**Review by stof — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: fixed

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: not sure.. https://github.com/symfony/symfony/blob/v3.2.8/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L577

but this definitely fixes tests :)

Comment (ro0NL) [inline src/Symfony/Component/DependencyInjection/Container.php]: i think `ContainerBuilder` needs to behave the same as `Container`, _after_ compilation. But that could be a separate PR i guess. The changelog doesnt mention `ContainerBuilder` really...

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: seems to break stuff :/

**Comment by nicolas-grekas** [comment]
I think we should move this to 3.3: it will trigger more notices in 3.2 apps and break some strict CI for no real benefit.

File: src/Symfony/Component/DependencyInjection/Container.php (hunk 2 above)
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Container.php]: well, once compiled, the array is not empty anymore, but we *do* want the deprecation in this case. So please remove the extra check you added there (which was not there before moving the notice btw, so you are the one adding a difference in the behavior)

**Comment by ro0NL** [comment]
3.2 is truly buggy though.. (@stof's comment in previous PR) i think the `ContainerBuilder` check will prevent those unexpected notices actually; while fixing `Container`.

otherwise, could you just commit the changelog against 3.2 :)) but you probably prefer a separate PR for that :P

**Comment by stof** [comment]
@ro0NL the ContainerBuilder check does not solve anything in 3.2.
The issue was that we were missing deprecations when getting a private service which was already instantiated previously (due to being a dependency of another service you got earlier). This happens both for a ContainerBuilder and the dumped container.
Your check fixes the testsuite only because the testsuite does not do any such bad `get` calls on a dumped container (as most tests don't bother about dumping the container and then loading the dumped container to run assertions)

**Comment by ro0NL** [comment]
@stof im still a bit confused.. (ive reverted the containerbuilder checks).

moving the private check up (above checking for shared services), truly fixes your comments from #22801, i.e. https://github.com/symfony/symfony/pull/22866/files#diff-8aeba2b88d0347062017ee160a6beb49R408 fails otherwise.

imo. it needs a fix for 3.2 (properly deprecating private services), but im missing something regarding the container builder. AFAIK the issue is `ContainerBuilder::get()` => `parent::get()`.

Updating [this](https://github.com/symfony/symfony/pull/22866/files#diff-00bc5ea153482178966f6617538a8e2eR217) call to `get()` still triggers the deprecation for private service `twig.extension.debug.stopwatch`. Im curious whats the root issue here, or at least what the proper fix is for either 3.2 or 3.3.

**Comment by ro0NL** [comment]
Think this approach https://github.com/symfony/symfony/blob/v3.2.8/src/Symfony/Component/DependencyInjection/Dumper/PhpDumper.php#L1488 needs to be done in container builder after compilation... ie. bypasssing `get()` internally for privates.

**Comment by nicolas-grekas** [comment]
My 2 cts: `ContainerBuilder` should not care about private or not. Thus, calling `parent::get()` shouldn't matter because `$this->privates` should not be populated.
I don't know where things fail right now but if we stick to this policy and tests fail, then I'd say these tests are broken, nothing else.

**Comment by nicolas-grekas** [comment]
For me, https://github.com/symfony/symfony/blob/7fbda2eb6dbff31b864d5fff4ceb42c2c87857b3/src/Symfony/Component/DependencyInjection/ContainerBuilder.php#L577 should be removed, that's the bug.
ContainerBuilder is a bootstrap-stage thing, and private/public is a "compiled-container" only thing.

**Comment by ro0NL** [comment]
Agree. I figured it out ;) i tend to qualify it a bug though.. as my perception is `compiled containerbuilder === compiled container === dumped  compiled container`, but for now we're good.

would be nice if `ContainerBuilder::compile()` returns a new (prepped) `Container`, that would allow to even further simplify (killing `isCompiled` and such), but that's quite a step though.

Squahsed&rebased. Ready for review :)

Sorry for the inconvenience, but this should do.

**Comment by stof** [comment]
@nicolas-grekas currently, you can keep using a ContainerBuilder after compilation (which is what some projects do as they don't dump the container to the cache, for instance Behat)

**Comment by ro0NL** [comment]
may i recall #20643 :) as containerbuilder is open for anything keeping privates empty is in line with that, and so a bugfix.

but the behavior is still wrong, however currently it's half right, half wrong.

File: src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/Tests/DependencyInjection/TwigExtensionTest.php]: reverted the whole change, as with privates being empty it keeps working :)

this test should fail whenever we decide to change behavior.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by stof** [comment]
We discussed this with @nicolas-grekas and here is the summary of changes we would want to get:

- add a hidden argument in `getServiceIds` to allow getting only public ids (i.e. opting in for the new behavior), and trigger a deprecation notice when getting the list with private ids in it (we have some existing code adding similar hidden arguments)
- retarget the PR to 3.3 (due to the above previous deprecation)
- try to keep the existing code filling the `privates` array in the ContainerBuilder, and debug test failures instead. As the 3.4 branch will allow using 4.0 deps, accessing private services in tests will be forbidden anyway. I suspect several failures are related to `getServiceIds` and so could be solved by the first point.

@ro0NL can you work on it ? Otherwise, I can take this PR over this evening to finish it

**Comment by ro0NL** [comment]
@stof i will finish it tonight :) so no changes for 3.2 right?

> try to keep the existing code filling the privates array in the ContainerBuilder

so we dont allow fetching privates using containerbuilder right (after compiling). `getServiceIds` will be the least problem :) it needs to bypass `get()` for references etc. internally. ie. what the phpdumper does.

**Comment by ro0NL** [comment]
@stof `Allow edits from maintainers` is enabled.. so feel free to add commits though :)

**Comment by fabpot** [comment]
Thank you @ro0NL.

**Comment by nicolas-grekas** [comment]
3.2 now merged into 3.3 so that next steps can happen on 3.3

**Pull request merged**